docs/source/notes/extending.rst

```diff
@@ -854,6 +853,62 @@ useful to confirm that a function that isn't present in the dictionary returned
 by ``get_overridable_functions`` cannot be overriden.
 
 
+.. _extending-torch-c++:
+
+Extending :mod:`torch` native API
+---------------------------------
+
+While ``__torch_function__`` allows to effectively extend PyTorch's pure python
+components' behavior, it does not allow to extend the parts of
+PyTorch implemented in C++. To that end, a :class:`Tensor` subclass can also
+define ``__torch_dispatch__`` which will be able to override the behavior at the
+c++ level.
+
+To effectively use this feature, it is important to know how the native part of
+PyTorch is implemented. The most important component there is what we call the
+"dispatcher" (the best description can be found in this [blog post](http://blog.ezyang.com/2020/09/lets-talk-about-the-pytorch-dispatcher/) even though it is slightly outdated).As
+hinted by its name, it is responsible to call the right backend
+function for a specific call of a function. For example, when calling
+``torch.add(a, b)``, the dispatcher will inspect both argument, figure out which
+"feature" (autograd, autocast, functionalization, etc) and which "backend" (CPU,
+CUDA, MPS, etc) should be used for this specific call and finally call all the
+right kernels.
+A very common thing done by a kernel is "redispatch". For example when running your
+neural network on GPU with autocast, the call will be first the autocast kernel that
+will handle any potential autocast logic and redispatch down. The next feature in line
+will be autograd that will properly create the autograd graph and redispatch down.
+Finally, we reach the backend kernel for CUDA which will launch the right CUDA kernel
+and return the final result. On the way out, autograd will attach the graph to the
+output and then autocast will have a chance to do any update it needs on exit.
+
+One of the configuration of the dispatcher is the order in which all these feature and backend keys are called. The latest list can be found in ``DispatchKey.h`` inside the ``DispatchKey`` enum. For the purpose of extending torch, the important piece of ordering is: vmap -> Autocast -> Autograd -> ZeroTensor -> Neg/Conj -> Functionalize -> Python -> Backends. The most important key for the purpose of this discussion is the ``Python`` one as every Tensor subclass with the ``__torch_dispatch__`` method defined will have the flag to call into this feature. It is from there that the user-defined method is called and where the behavior can be overwritten arbitrarily. From there, calling the provided ``func`` again will perform a "redispatch".
```

Comment:
```suggestion
One configuration of the dispatcher is the order in which all these feature and backend keys are called. The latest list can be found in ``DispatchKey.h`` inside the ``DispatchKey`` enum. For the purpose of extending torch, the important ordering to observe is: vmap -> Autocast -> Autograd -> ZeroTensor -> Neg/Conj -> Functionalize -> Python -> Backends. The most important key for the purpose of this discussion is the ``Python`` one as every Tensor subclass with the ``__torch_dispatch__`` method defined will have the flag to call into this feature. It is from there that the user-defined method is called and where the behavior can be overwritten arbitrarily. From there, calling the provided ``func`` again will perform a "redispatch".
```
I get the sense of what you’re talking about, but I don’t know what “flag” means.

Reply:
"flag" is not a great name indeed. Removing to just say that we call it.